app/routes/atlas/reduceDimensions.ts

```diff
@@ -0,0 +1,59 @@
+import { TSNE } from '@keckelt/tsne'
+import { UMAP } from 'umap-js'
+
+export type Point = {
+  x: number
+  y: number
+}
+
+export function reduceDimensions(data: number[][], alg: 'tsne' | 'umap') {
+  // https://youtu.be/NEaUSP4YerM
+  if (alg === 'tsne') {
+    let tsne = new TSNE({
+      epsilon: 10, // learning rate (10 = default)
+      perplexity: 30, // how many neighbors does each point influence  (30 = default)
+      dim: 2, // dimensionality of the embedding (2 = default)
+    })
+    tsne.initDataRaw(data)
+    let iterations = 100
+
+    for (let i = 0; i < iterations; i++) {
+      tsne.step()
+    }
+
+    return tsne.getSolution()
+  }
+
+  // https://youtu.be/eN0wFzBA4Sc
+  if (alg === 'umap') {
+    let umap = new UMAP()
+    return umap.fit(data)
+  }
+
+  return []
+}
+
+type GridSize = { width: number; height: number }
+
+let defaultGridSize = { width: 10, height: 10 }
+
+export function toGridIndices(
```

Comment:
This approach has the disadvantage of leaving empty space in the grid.

To do: write a better grid conversion function. Possibly good reference:
https://github.com/kylemcdonald/CloudToGrid/blob/master/CloudToGrid.ipynb